[BUGFIX] Not working properly on self-hosted runners

thommyhh opened the pull request:
* remove fixed default auth socket path
* call `ssh-agent` with auth socket path, if not explicitly given,
  to let the system choose the path
* read SSH_AUTH_SOCKET and SSH_AGENT_PID from `ssh-agent` command output
  and export them as variables
* update `dist/index.js`
* update readme with section about exported variables

Fixes #16

rbellamy commented:
Maybe I'm reading this incorrectly, but wouldn't this kill the agent for other runners on the same machine?

rbellamy commented:
Adding the run ID seems like a wise choice to ensure the agent is using a unique `AUTH_SOCK`:

https://github.com/webfactory/ssh-agent/issues/16#issuecomment-580482355

thommyhh commented:
@rbellamy As far as I know, `ssh-agent` picks a random unique file name for the auth socket, if none is provided. This is why I removed the default. By exporting the process ID `$SSH_AGENT_PID`, we can kill that one agent. No other agents are killed.

mpdude commented on README.md:
```diff
@@ -57,6 +57,20 @@ The `ssh-agent` will load all of the keys and try each one in order when establi
+## Exported variables
+The action exports `SSH_AUTH_SOCK` and `SSH_AGENT_PID` through the Github Actions core module.
+The `$SSH_AUTH_SOCK` is used by several applications like git or rsync to connect to the SSH authentication agent.
+The `$SSH_AGENT_PID` can be used to kill the SSH agent, e.g. when using on a self-hosted runner:
```
I'd shorten this to only say "... contains the process id of the agent running for this job". I think it should be the action's responsibility to kill the agent.

thommyhh replied:
I added a post action to kill the agent. Seems to work as expected.

mpdude commented on dist/index.js:
```diff
@@ -62,7 +62,7 @@ try {
     const homeSsh = home + '/.ssh';
 
-    const privateKey = core.getInput('ssh-private-key').trim();
+    const privateKey = core.getInput('ssh-private-key');
```
Could you rebase this PR against `master` please? Then this change should go away.

mpdude commented on dist/index.js:
```diff
@@ -71,14 +71,27 @@ try {
 
     console.log(`Adding GitHub.com keys to ${homeSsh}/known_hosts`);
-    fs.mkdirSync(homeSsh, { recursive: true});
+    fs.mkdirSync(homeSsh, { recursive: true });
```
Same as before, should go away with a rebase.

mpdude commented on dist/index.js:
```diff
@@ -71,14 +71,27 @@ try {
+    if (authSock && authSock.length > 0) {
+        sshAgentOutput = child_process.execFileSync('ssh-agent', ['-a', authSock]);
+    } else {
+        sshAgentOutput = child_process.execFileSync('ssh-agent')
```
This is a slight BC break since now _not_ providing a value for the `ssh-auth-sock` input will bind it to a random port, instead of the fixed value we used previously. (Yes, I completely understand that this is part of the goal to be able to run multiple instances in parallel; just sayin').

mpdude replied:
Since we're still in the 0.x version ranges, IMO nothing we should worry about too much.

thommyhh replied:
It binds to a random socket file, just for being precise. I don't think the value matters as long as it written to the `$SSH_AUTH_SOCK` variable.

thommyhh commented:
@mpdude I updated the branch with the latest changes from master, but the changes did not disappear. I don't know, how to trigger GitHub to re-read the base branch (tried changing base).

I also added the post action to kill the agent automatically, as suggested. I wondered, if there is something like a post job script, to clean things up.

zhigang1992 commented:
I think you can merge master and then push?

thommyhh commented:
@zhigang1992 
That's what I did, but GitHub did not update the base branch anyway.

mpdude commented:
I have released 0.3.0 with this change. 

Thank you @thommyhh and everyone involved!

mkrakowitzer commented:
Thanks @thommyhh @mpdude for resolving this!
